### ISSUE TITLE
Some albums covers do not appear with DeaDBeeF beta3

### ISSUE THREAD

**Issue by Npa972** [DeaDBeeF-Player/deadbeef#2740]
Install DeaDBeeF beta 3 on Arch

Some covers appears and many others do not. I think only covers in metadata appears but not the one stored in .cache/deadbeef/covers.

I have no problem with 1.8.8 version. All the covers appears (i have 829 folders in .cache/deadbeef/covers/)

Thank you.

**Comment by Alexey-Yakovenko** [comment]
Hi,

> I have no problem with 1.8.8 version. All the covers appears (i have 829 folders in .cache/deadbeef/covers/)

deadbeef 1.9.0 comes with a new artwork plugin, which uses different cache format, different cache location, and different caching mechanisms. So, in the context of this issue -- this folder can be ignored.

I will need some additional information:

Does the music file in question come with album art?
What is the file format?
What is the image format?
Is the image embedded in the file, or is located on disk next to the audio file?
Is it using one of the file name patterns, configured in the artwork plugin?
Did you check, that the artwork plugin has all the necessary options enabled?

**Comment by Npa972** [comment]
Thanks for your answer. I need time to check some points you ask for.

**Comment by Npa972** [comment]
I have all my music in only one folder (19627 titles and 1484 albums).
I've made a fresh install with beta3 version (I deleted cache and config of 1.8.8).
The covers now appears in folder .cache/deadbeef/covers2/
If a cover is missing, there is a file "artist-album.jpeg" but it's a text file with 0 octet.
It was the same with 1.8.8 but i could replace the text file by a image with exactly the same name and the cover appears when update cover in the play list.
But now the jpg file always return to a text file so i can't add covers.
Another problem : i can only sort the play list with the menu. Clicking in the field on top of the play list (genre, album, artist...) do not work.
Sorry i'm not very technical just a simple user
![Capture d’écran_2022-02-08_19-44-51](https://user-images.githubusercontent.com/49346431/153095753-12bdfa61-b84b-4ebb-97b2-599ec2d156cf.png)
![Capture d’écran_2022-02-08_19-52-30](https://user-images.githubusercontent.com/49346431/153095786-f3f467b3-256f-470e-91b1-e1a3c2908f3e.png)
![Capture d’écran_2022-02-08_19-55-12](https://user-images.githubusercontent.com/49346431/153095808-5a241a9e-f1a4-4111-b539-39a93e621c0e.png)
.

**Comment by Npa972** [comment]
One more thing group by album (%album%) is not kept by deadbeef after quit.

**Comment by Alexey-Yakovenko** [comment]
I wrote down both reported bugs, will fix them in beta4 -- thanks.

Regarding your album art problem...
As far as I understand, you're trying to manually place files to deadbeef's cache, and then assume that deadbeef will use them in some way, but that's not really how things work.

The only solution that comes to mind for your use case is to embed the album art images into every audio file.

However, what you described also doesn't quite seem right -- deadbeef should not be overwriting the marker files on each launch. This may indicate another issue, which I will have to check.

**Comment by Alexey-Yakovenko** [comment]
@Npa972 I discovered an unfortunate bug, which seems to have been introduced right before I released the beta3.
I don't know whether it's directly related to your problem or not, but it may be.
Can I ask you to try this build and see if it changes anything for the better?
https://sourceforge.net/projects/deadbeef/files/travis/linux/master/deadbeef-static_devel-1_x86_64.tar.bz2/download

**Comment by Npa972** [comment]
Sorting is now working when clicking on genre, album...
Grouping by %album% is not kept if quit and restart.
And there is a problem - for me - with album art : Very slow to appear on each start, impossible to add image to cache.
By comparison if you go quickly from top to bottom of the playlist (mine is important) in 1.8.8, image are here instantly. It's the same on start.
In the devel version or beta3 the behaviour is strange : if you go top to bottom / bottom to top, images disappears and it takes a very long time to reappear.

**Comment by Npa972** [comment]
One more thing when moving thru the play list, images disappears no matter they are embedded or download to cache with lastfm.
And DeadBeef can freeze when moving quickly into the play list
![Capture d’écran_2022-02-09_16-06-16](https://user-images.githubusercontent.com/49346431/153281752-776232e5-8cc8-44fd-a03c-11f83ac9b8ea.png)
.

**Comment by Alexey-Yakovenko** [comment]
> And there is a problem - for me - with album art : Very slow to appear on each start, impossible to add image to cache.

what computer specs do you have? I'm on a 2017 i5, and my audio files are stored on an external USB HDD. I'm not experiencing significant slowness with album art, and I tried placing the image in the cache by hand which worked just fine.

You mentioned that you're getting deadbeef UI to freeze while scrolling -- can you sample the process in GDB to help investigate it?

**Comment by Alexey-Yakovenko** [comment]
ok, I just reproduced the freeze. It occurred after deleting the cache, and slowly scrolling through playlist.

**Comment by Npa972** [comment]
Thank you very much it's better.
Grouping by album is working (stop an start DeadBeef)
I spent some hours to add all the missing album's image but they are all here ! (i have backup up the cache directory).
But image still appears slowly after starting, if moving slowly into the playlist (for example searching a specific album). The memory consumption is high during this (processor also). Memory start à 80 Mb and finish at more than 300. With 1.8.8 memory never goes as high (95 Mb perhaps...).
Is there a problem with my album artwork plugin settings ?
But no freeze.
![Capture d’écran_2022-02-12_16-38-07](https://user-images.githubusercontent.com/49346431/153727798-5c53641a-0e36-49ea-b9f0-4562ae51f615.png)
My computer is a recent i3 (music on usb HDD on this computer).
![Capture d’écran_2022-02-12_16-54-01](https://user-images.githubusercontent.com/49346431/153728068-400e75e1-8df3-4625-8aef-b4e162c1b07f.png)

**Comment by Alexey-Yakovenko** [comment]
indeed, when loading from tags / local folders is enabled -- it can be significantly slower, than loading from cache.
This was done on the assumption that deadbeef will be fast enough to load the images from local places without any caching mechanism, but seems like this was a mistake.
Just checking that an audio file doesn't have embedded cover can take 2 seconds on my machine.

I will try to change the code to use the cache for all scenarios.

In the meantime, to get the max speed, it's recommended to disable "Fetch from embedded tags".

**Comment by Alexey-Yakovenko** [comment]
I applied the change, and now the disk cache will be used for all album cover finding mechanisms (music folder + embedded + web), and performance has improved quite drastically.
If you want to help testing it -- it would be greatly appreciated.
The build is here: https://sourceforge.net/projects/deadbeef/files/travis/linux/master/deadbeef-static_devel-1_x86_64.tar.bz2/download

**Comment by Npa972** [comment]
I try the last build and all the covers are in cache now (124 Mb, 1492 files). So i could disable everything in artwork plugin (music folder + embedded + web).
I backup the cache and use it for a fresh install on another computer (celeron) and it works perfectly.
The performance are very good (covers appears instantly on both computers).
Just one remark, the memory consumption is high (260 MB and even more when "embedded" was enable) when moving slowly into the entire play list (witch is long in my case). I think that the 1.8.8 never reached more than 100 Mb for this task.
![Capture d’écran_2022-02-13_07-58-37](https://user-images.githubusercontent.com/49346431/153752877-b2d41825-9a7d-40e3-8f53-a0598409c952.png)

**Comment by Alexey-Yakovenko** [comment]
Thanks a lot for the feedback. Not it seems like the functionality is ok, I will start looking into memory consumption

**Comment by Alexey-Yakovenko** [comment]
In my tests, RAM consumption is in the acceptable margins -- it normally stays around ~110MB.
This is ~20MB difference compared to when album art is OFF.
I'm running artwork plugin with default settings, which means it loads artwork images scaled to 256x256px.

Given that internal RAM cache is 50 images, we can do a simple calculation:
256*256*4*50 = ~13MB

So this sort of makes sense.

Are you sure that you didn't change the artwork image size setting to something larger?

**Comment by Alexey-Yakovenko** [comment]
another test: ubuntu 20.04 / GTK2
album art off: 55MB
album art on, default settings + lastfm ON, RAM cache full: gets to 134MB

still looks acceptable to me.

**Comment by Npa972** [comment]
Yes i kept default setting (256).
I moved the entire playlist to load all images and i reached 254 MB but perhaps it's normal with 1492 images in cache ?
From what i what i understand with your calculation (256*256*4*50), the number of images count in Ram consumption ?
Anyway it's not very important as a user because there is no effect on performance and i never browse the entire playlist like this it's too long...
Thanks again for your work.

**Comment by Alexey-Yakovenko** [comment]
no, it's not normal, since RAM cache is limited to 50 images.

**Comment by Alexey-Yakovenko** [comment]
> From what i understand with your calculation (`256*256*4*50`), the number of images count in Ram consumption ?

Yes. It's the projected minimal RAM consumption. It's always significantly larger than this calculation -- e.g. twice as much is normal. But in your case, it seems like it's 10 times larger than it should be.

I would like to hear how much RAM deadbeef takes on your machine when album art is off

**Comment by Alexey-Yakovenko** [comment]
I agree that it's not super important, if it doesn't affect the user experience, but I'm sure there will be more users who would file bug reports about this increased RAM consumption after I release 1.9.0 -- so I'd rather prefer to fix this while I can :)

**Comment by Npa972** [comment]
With album art off (i just delete the column album art) i had 75 MB after browsing all the playlist.
Ram consumption do not change (74 MB when starting Deadbeef and 75.4 MB after browsing entire playlist).

**Comment by Npa972** [comment]
Album art ON and moving top to bottom slowly enough to load images i get 264 MB

**Comment by Alexey-Yakovenko** [comment]
thanks for the information. I'll keep looking for any clues that may be causing this. will let you know if I find anything.

**Comment by Alexey-Yakovenko** [comment]
A few more questions:

* Can you try measuring the RAM level using htop instead of the app you're currently using? Just so that we're aligned / use the same tools / talking same numbers?
* And.. when you're saying the RAM consumption goes to 264MB -- is it permanent, never dropping back to ~110-130?

**Comment by Alexey-Yakovenko** [comment]
I just tried again, 400 cover images totalling 123MB in the disk cache, slowly scrolled through 2 large playlist which use them all -- RAM consumption 94.4MB (and that's a debug build)

**Comment by Npa972** [comment]
244 MB with htop
I'm waiting to see if it's dropping back (no change after 10 mn).
![Capture d’écran_2022-02-13_16-05-51](https://user-images.githubusercontent.com/49346431/153772758-ef22e76a-8988-4331-9a39-965a7d2b547e.png)
k

**Comment by Alexey-Yakovenko** [comment]
ok, thanks. I asked people on discord to try to reproduce this. so far, my test environments are not enough.

**Comment by Alexey-Yakovenko** [comment]
So.. I still wasn't able to find the large memory consumption bug that you're experiencing, even after installing and trying it on arch linux (I used Manjaro though).

However, while digging I found another related memory spender, and fixed it.

I don't know whether it will affect your situation, but it will definitely reduce the memory consumption when starting from clean disk cache.

**Comment by Npa972** [comment]
I tried the last build on two computers (one with clean disk cache). Both archlinux up to date, one i3 and one celeron.
Moving top bottom bottom top to load images gave me 245 MB on both computer. It's better than before.
Setting is 256.
music folder + embedded + web are disable in artwork plugin.
1488 covers in cache for 186 MB.
Apparently it is not dropping back.
One question : i have some "heavy" images (50 are more than 500 kb and 14 more than 1 MB) ; could it be a problem ?

**Comment by Alexey-Yakovenko** [comment]
> One question : i have some "heavy" images (50 are more than 500 kb and 14 more than 1 MB) ; could it be a problem ?

maybe, but I already tried to re-create this condition, and nothing changed

**Comment by Alexey-Yakovenko** [comment]
would it be an option for you to send me your playlist (dbpl file) + your album art cache in a zip file, via some file sharing like google drive? preferrably over email

**Comment by Npa972** [comment]
Thank you !

https://drive.google.com/file/d/1GFSCcnHmlxunfz1_paKq0ptrhCeF3MAe/view?usp=sharing

https://drive.google.com/file/d/1oZYSt7WYynzbyjAiSOJxPUXCoJU7hb0J/view?usp=sharing

--

**Comment by Alexey-Yakovenko** [comment]
thanks, downloaded.

**Comment by Alexey-Yakovenko** [comment]
@Npa972 FYI: I have changed the format in which files are named in the disk cache.
However, I added an option to use the "artist - album.jpg" format for the people who want that.
It's very ambiguous and limiting, but I recognize your special needs, and if that's what you want -- you can enable the option in the artwork plugin preferences called "Simplified cache file names".

**Comment by Alexey-Yakovenko** [comment]
I was able to reproduce the problem with your playlist / cache. I think I'm getting some ideas where to look next, thanks for the data!

**Comment by Npa972** [comment]
I don't understand some points.
The format "artist-album.jpg" was loaded by Deadbeef when creating the covers (web and embedded), i didn't do anything for that. Is it because all my music is in a single folder ?
And what is special need compare to normal needs ? I don't know what i want particularly compared to a standard Deadbeef user ?
I try the last build (clean cache with "simplified cache files names" enable) and later i will try with a new clean cache and "simplified cache file name" disable to see the difference.
And... thanks you again.

**Comment by Alexey-Yakovenko** [comment]
With the help of your playlist and the cache, I found what was consuming all that RAM, and removed it. There's still some weird RAM consumption, but not nearly as much as it was. Please let me know how it works for you :)

> The format "artist-album.jpg" was loaded by Deadbeef when creating the covers (web and embedded), i didn't do anything for that. Is it because all my music is in a single folder ?

It was probably me misunderstanding how you are using the deadbeef's artwork cache. You mentioned that you're putting the files there by hand etc.
If you are not doing this -- then there's nothing to worry about.

**Comment by Alexey-Yakovenko** [comment]
> And what is special need compare to normal needs ? I don't know what i want particularly compared to a standard Deadbeef user ?

There's nothing special in your use-case. My goal is to make sure that album art works correctly regardless of user preferences, ambiguous metadata, various folder structures, all files in the same folder, etc.

**Comment by Npa972** [comment]
OK i understand.

I add files by and only if they are missing after "scanning" embedded an web.

**Comment by Alexey-Yakovenko** [comment]
so, well, this is a quite special use case :) with the new less-ambiguous cache filenames it will be much harder to create a proper filename by hand -- that's why I decided to keep the simplified option.
(also, I will probably change this new naming scheme a few times before shipping, since it may need adjustments if any issues are identified)

**Comment by Npa972** [comment]
Hi, i did a lot of testing with 1.9 RC1 and i really appreciate the simplified option. It's easy to add covert art by hand (approximately 140 albums on a total of 1500) when nothing else works.
But, for me, the memory problem, related with covers, is not resolved. With all the covers in cache, the behavior of Deadbeef is "strange" compared to 1.8.8. It's looks like covers are always charging / refreshing when moving into the play list.
I tried several combination with fresh config (uncheck simplified, not adding any covers by hands...) but the result is the same.
![Capture d’écran_2022-02-27_21-56-34](https://user-images.githubusercontent.com/49346431/155912590-fb73419e-a62c-43d7-a563-6eeb9f4a1ce6.png)
![Capture d’écran_2022-02-27_22-07-41](https://user-images.githubusercontent.com/49346431/155912602-6b1200a1-4686-43da-bc6c-7726754c1947.png)

**Comment by Alexey-Yakovenko** [comment]
@Npa972 This memory problem has diverged a lot from the original reported bug, so for your convenience I've created a new bug: https://github.com/DeaDBeeF-Player/deadbeef/issues/2762
